### PR TITLE
Action Plan deprication initial commit

### PR DESCRIPTION
**Pull request by akssinhaamit**
# Motivation and Context

# What has changed

# How to test?

# Links

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java [hunk 1]
```diff
@@ -82,38 +86,64 @@ public void acceptNotification(final CaseNotification notification) throws CTPEx
...
   }
 
-  private ActionCase createActionCase(CaseNotification notification) {
-    UUID actionPlanId = UUID.fromString(notification.getActionPlanId());
-    ActionPlan actionPlan = actionPlanRepo.findById(actionPlanId);
-    if (actionPlan == null) {
-      throw new IllegalStateException(
-          String.format(ACTION_PLAN_NOT_FOUND, actionPlanId.toString()));
+  private ActionCase createActionCase(CaseNotification notification) throws NullPointerException {
+    if (notification.getActionPlanId() != null) {
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java:90]: Could we have a comment here to remind us to remove this code once we get rid of Action Plans?

File: src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java [hunk 2]
```diff
@@ -126,24 +156,31 @@ private void saveActionCase(ActionCase actionCase) {
...
       log.with("case_id", actionCaseId.toString()).error("No case found to update");
       throw new IllegalStateException(
           String.format(ACTION_CASE_NOT_FOUND, actionCaseId.toString()));
     }
-    ActionPlan actionPlan = actionPlanRepo.findById(actionPlanId);
-    if (actionPlan == null) {
-      log.with("action_plan_id", actionPlanId.toString()).error("No action plan found");
-      throw new IllegalStateException(
-          String.format(ACTION_PLAN_NOT_FOUND, actionPlanId.toString()));
+    if (actionPlanId != null) {
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java:167]: Likewise, would be good to get a comment to make it easier to tear this code out after

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+    }
+    log.debug("Getting Email cases against collectionExerciseId and event active enrolment");
+    List<ActionCase> letter_cases =
+        actionCaseRepository.findByCollectionExerciseIdAndActiveEnrolment(
+            collectionExerciseId, false);
+    if (letter_cases.size() > 0) {
+      ActionTemplate actionTemplate = actionTemplateService.mapEventTagToTemplate(eventTag, false);
+      if (actionTemplate != null) {
+        log.with("Letter cases", letter_cases.size()).info("Processing letter cases");
+        processLetterCases(letter_cases, actionTemplate, survey, collectionExercise, instant, true);
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:132]: It's a little weird that we filter email_cases in this method but filter letter_cases in the method we're calling. It'd be good to have some consistency; either stream both types of case here, or stream both types of case in processLetterCases/processEventCases.

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+            + getExerciseRefWithoutSurveyRef(collectionExercise.getExerciseRef());
+    final String now =
+        DateTimeFormatter.ofPattern("ddMMyyyy_HHmm")
+            .withZone(ZoneId.systemDefault())
+            .format(instant);
+    String filename = String.format("%s_%s.csv", fileNamePrefix, now);
+    log.info("filename: " + filename + ", uploading file");
+    log.with("actionType", actionTemplate.getType()).info("Processing Print File");
+    boolean isSuccess = printFileService.processPrintFile(filename, letterEntries);
+    log.with("process file status?", isSuccess)
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:186]: Would reword this to "file processed?" just for a little more clarity when reading logs

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+          templateHandler,
+          isSuccess,
+          collectionExercise.getId(),
+          survey.getId(),
+          instant);
+    } else {
+      updateCaseActionEvent(actionEvent, instant);
+    }
+  }
+
```
Comment (mfcrocker) [inline src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:496]: It feels like a lot of the methods below belong in other classes, eg setParties feels like it should be part of ActionCaseParty or the party svc client. This is a stylistic choice, but it's worth having a think about where they most logically fit.

File: src/main/resources/database/changes/release-24/changelog.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+databaseChangeLog:
+
+  - changeSet:
+      id: 24-1
+      author: Amit Sinha
+      changes:
+        - sqlFile:
+            comment: active Enrolment addition
```
Comment (mfcrocker) [inline src/main/resources/database/changes/release-24/changelog.yml:8]: It does a little bit more than just adding the active enrolment - could we get a fuller description here please?

**Review by mfcrocker — CHANGES_REQUESTED** [review]
In general this looks good! I've got some suggestions that are optional, but I need the cron stuff checking and the 0-indexed nudge emails changing before approving this

File: _infra/helm/action/values.yaml [hunk 7]
```diff
@@ -31,19 +31,19 @@ service:
 crons:
   distributionScheduler:
     name: action-distribution-scheduler
-    cron: "* * * * *"
+    cron: "0 0 0 * * ? 2030"
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply _infra/helm/action/values.yaml]: I changed it as soon as I realised. Not sure why it did not get refreshed but now its `0 0 1 12 *`. Hope its ok!

File: src/main/java/uk/gov/ons/ctp/response/action/endpoint/ActionEventEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  @RequestMapping(
+      value = "/process-event",
+      method = RequestMethod.POST,
+      consumes = "application/json")
+  public ResponseEntity processEvents(@RequestBody @Valid Event event) {
+    log.with("collectionExercise", event.getCollectionExerciseID())
+        .with("EventTag", event.getTag())
+        .info("Processing Event");
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/endpoint/ActionEventEndpoint.java:33]: Don't think it will be noisy, but please correct me if I am wrong. It will only be called once for each event of a single collection exercise.

File: src/main/java/uk/gov/ons/ctp/response/action/representation/events/Event.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Event {
+  public enum EventTag {
+    mps,
+    go_live,
+    reminder,
+    reminder2,
+    reminder3,
+    nudge_email_0,
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/representation/events/Event.java:22]: I do understand, but this is how the collection exercise implementation of nudge email is done. This entry is just to match the incoming event name. If it needs changing the collection exercise svc nudge email implementation will have to change :-( .

File: src/main/java/uk/gov/ons/ctp/response/action/service/ActionTemplateService.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  /** Event Tag to ActionTemplate mapper */
+  private enum TemplateMapper {
+    mps("mps"),
+    go_live("go_live"),
+    reminder("reminder"),
+    reminder2("reminder"),
+    reminder3("reminder"),
+    nudge_email_0("nudge"),
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ActionTemplateService.java:46]: I do understand, but this is how the collection exercise implementation of nudge email is done. This entry is just to match the incoming event name. If it needs changing the collection exercise svc nudge email implementation will have to change :-( .

File: src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java (hunk 1 above)
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java:90]: Sure 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java (hunk 2 above)
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/CaseNotificationService.java:167]: Sure 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java (hunk 3 above)
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:132]: yes sure it can be consistent. Will extract it as a method.

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java (hunk 4 above)
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:186]: 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java (hunk 5 above)
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:496]: Sure, you are absolutely right. This does require further refactoring. Can work on this for sure

File: src/main/resources/database/changes/release-24/changelog.yml (hunk 6 above)
Comment (akssinhaamit) [reply src/main/resources/database/changes/release-24/changelog.yml:8]: 👍

**Review by mfcrocker — APPROVED** [review]

File: src/main/java/uk/gov/ons/ctp/response/action/printfile/LetterEntry.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import uk.gov.ons.ctp.response.action.representation.ActionTemplateDTO.Handler;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PUBLIC)
+public class LetterEntry {
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/response/action/printfile/LetterEntry.java:13]: Could we re-use PrintFileEntry rather than have a new class?

They look pretty much the same and I prefer the print file terminology

**Review by warrenbailey — CHANGES_REQUESTED** [review]
a few minor things, while I test out the changes

File: src/main/java/uk/gov/ons/ctp/response/action/printfile/LetterEntry.java (hunk 11 above)
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/printfile/LetterEntry.java:13]: The reason for a seperate one is to make the deletion easier. Also to make sure that the new changes doesn't interact with anything which is being used by the previous version. Which will still be used till the time we create the migration script.

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 12]
```diff
@@ -0,0 +1,774 @@
...
+            .handler(handler)
+            .status(
+                status
+                    ? ActionEvent.ActionEventStatus.PROCESSED
+                    : ActionEvent.ActionEventStatus.FAILED)
+            .collectionExerciseId(collectionExerciseId)
+            .surveyId(UUID.fromString(surveyId))
+            .processedTimestamp(status ? Timestamp.from(instant) : null)
+            .build();
+    actionEventRepository.save(actionEvent);
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:250]: Its one at a time. we don't actually want to roll back the entire transaction if any one fails do we?

File: src/main/java/uk/gov/ons/ctp/response/action/service/NotifyEmailService.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import uk.gov.ons.ctp.response.action.config.AppConfig;
+
+@Service
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/NotifyEmailService.java:17]: 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/NotifyLetterService.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+import uk.gov.ons.ctp.response.action.config.AppConfig;
+import uk.gov.ons.ctp.response.action.message.UploadObjectGCS;
+import uk.gov.ons.ctp.response.action.message.instruction.ActionRequest;
+import uk.gov.ons.ctp.response.action.printfile.Contact;
+import uk.gov.ons.ctp.response.action.printfile.LetterEntry;
+import uk.gov.ons.ctp.response.action.printfile.PrintFileEntry;
+
+@Log4j
+@Service
+public class NotifyLetterService {
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/NotifyLetterService.java:26]: 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/NotifyLetterService.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    return success;
+  }
+
+  private String createJson(List<LetterEntry> printFile) throws JsonProcessingException {
+    ObjectMapper mapper = new ObjectMapper();
+    return mapper.writeValueAsString(printFile);
+  }
+
+  protected List<PrintFileEntry> convertToPrintFile(List<ActionRequest> actionRequests) {
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/NotifyLetterService.java:77]: Copy paste😁. Changed 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 16]
```diff
@@ -0,0 +1,774 @@
...
+        .with("event_tag", eventTag)
+        .info("Started processing");
+    Instant instant = Instant.now();
+    CollectionExerciseDTO collectionExercise = getCollectionExercise(collectionExerciseId);
+    SurveyDTO survey = getSurvey(collectionExercise.getSurveyId());
+    log.debug("Getting Email cases against collectionExerciseId and event active enrolment");
+    List<ActionCase> email_cases =
+        actionCaseRepository.findByCollectionExerciseIdAndActiveEnrolment(
+            collectionExerciseId, true);
+    if (email_cases.size() > 0) {
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:103]: 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 17]
```diff
@@ -0,0 +1,774 @@
...
+            .with("event", eventTag)
+            .error("No Template registered against the event and active enrolment");
+      }
+    } else {
+      log.with("collectionExercise", collectionExercise)
+          .with("event", eventTag)
+          .info("No Emails to process");
+    }
+    log.debug("Getting Email cases against collectionExerciseId and event active enrolment");
+    List<ActionCase> letter_cases =
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:120]: 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 18]
```diff
@@ -0,0 +1,774 @@
...
+        actionCaseRepository.findByCollectionExerciseIdAndActiveEnrolment(
+            collectionExerciseId, false);
+    if (letter_cases.size() > 0) {
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:123]: 👍

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 19]
```diff
@@ -0,0 +1,774 @@
...
+   * processes email cases if actionable Gets all letter cases against collectionExerciseID
+   * (isActiveEnrolment == false) Maps letter cases to ActionTemplate processes letter cases if
+   * actionable
+   *
+   * @param collectionExerciseId - is passed by collectionexercisesvc scheduler as a part of the
+   *     event process.
+   * @param eventTag - is passed by collectionexercisesvc scheduler as a part of the event process.
+   */
+  @Async
+  public void processEvents(UUID collectionExerciseId, String eventTag) {
```
[the remark this answers is not in the document]
Comment (akssinhaamit) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:92]: you might be right.. Although I avoided it not to bring two async process up at the same time. I might have to look into this further to find if there would be any performance issues.

[the remark this answers is not in the document]
Comment (warrenbailey) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java:92]: I actually meant refactor the contents of the methods, i.e. keep the async processEvent() method but call out to two private methods processEmailEvents and processLetterEvents. It just keeps the method size smaller then.

File: src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java [hunk 20]
```diff
@@ -0,0 +1,774 @@
...
+    log.with("collectionExerciseId", collectionExerciseId)
+        .with("event_tag", eventTag)
+        .info("Started processing");
+    Instant instant = Instant.now();
+    CollectionExerciseDTO collectionExercise = getCollectionExercise(collectionExerciseId);
+    SurveyDTO survey = getSurvey(collectionExercise.getSurveyId());
+    log.debug("Getting Email cases against collectionExerciseId and event active enrolment");
+    List<ActionCase> email_cases =
```
[the remark this answers is not in the document]
Comment (warrenbailey) [reply src/main/java/uk/gov/ons/ctp/response/action/service/ProcessEventService.java]: yeah jumping between languages is always fun

File: src/main/java/uk/gov/ons/ctp/response/action/printfile/LetterEntry.java (hunk 11 above)
Comment (warrenbailey) [reply src/main/java/uk/gov/ons/ctp/response/action/printfile/LetterEntry.java:13]: Ok that makes sense, I prefer the PrintFileEntry name as that what it is so maybe during the clean up we could rename it

**Review by warrenbailey — APPROVED** [review]
